include/boost/numeric/ublas/tensor/functions.hpp

```diff
@@ -229,7 +259,19 @@ namespace boost::numeric::ublas
 
         nc[m - 1] = nb[0];
 
-        auto c = result_tensor_t<value_type,decltype(nc),layout_type>(nc, value_type{});
+        using c_extents_type = std::decay_t< decltype(nc) >;
+
+        using t_engine = tensor_engine< 
+            c_extents_type,  
+            std::conditional_t< 
+                std::is_same_v< layout_type, first_order >,
+                layout::first_order<c_extents_type>,
+                layout::last_order<c_extents_type>
+            >,
+            rebind_storage_t<c_extents_type,array_type,value_type>
```

Comment:
why do we need to rebind here?

Reply:
Because storage type can be user-defined and I wanted to have a standard way to create result type in every function without thinking about anything.

Reply:
you have included some sort of type promotion. I would like to postpone the topic of type promotion.

Reply:
There is no type promotion. I'm using `rebind_storage_t` to have a standard way to select the correct container and in the case of static it just gets the new static container with modified size and correct `value_type`.

Reply:
This will be more useful in static prod because it might use static storage which after contraction needs to change in size so how we do it, there comes the use of `rebind_storage` as we don't want any increase in complexity inside the prod function. To be consistent with the static prod I added it to the dynamic prod functions which don't do any kind of promotion just act as a placeholder.

Use:
- static prod with static storage => rebinds the storage with the new size
- static prod with dynamic storage => does not do anything but returns the same storage. This reduces the complexity inside the prod functions.

Reply:
Let me phrase it differently. With `rebind_storage_t` the resulting storage type depends on the input storage type. In this case, the user of the library needs to look into `type_traits_tensor.hpp` for understanding the output type.  Moreover, if the user wants to only have statically allocated containers, this rule forces him/her to not use the `prod` function or to change the implementation of `rebind_storage_t`.

Therefore, we should not to have that feature (at least for now) and at most use [tag dispatching](https://www.boost.org/community/generic_programming.html). Again the `prod` function with that signature should only be implemented for containers with dynamic extents and strides.

Reply:
Changed it to `rebind_storage_size` for static products with is necessary for it otherwise there is no way to use it with static storage and removed `rebind_storage` which may be confusing instead I used storage traits to rebind in functions for transforming `conj` and 'img'.